coderd/database/dbauthz/dbauthz.go

```diff
@@ -5507,6 +5524,30 @@ func (q *querier) UpdateWorkspaceAgentLogOverflowByID(ctx context.Context, arg d
 }
 
 func (q *querier) UpdateWorkspaceAgentMetadata(ctx context.Context, arg database.UpdateWorkspaceAgentMetadataParams) error {
+	// Fast path: Check if we have an RBAC object in context.
+	// This is set by the workspace agent RPC handler to avoid the expensive
+	// GetWorkspaceByAgentID query for every metadata update.
+	//
+	// NOTE: The cached RBAC object is refreshed every 5 minutes. After a prebuild
+	// claim, there may be up to a 5-minute window where this uses stale owner_id.
+	if rbacObj, ok := WorkspaceRBACFromContext(ctx); ok {
+		// Verify the RBAC object is valid (has required fields).
+		if rbacObj.Owner != "" && rbacObj.OrgID != "" && rbacObj.Owner != uuid.Nil.String() && rbacObj.OrgID != uuid.Nil.String() {
+			act, ok := ActorFromContext(ctx)
+			if !ok {
+				return ErrNoActor
+			}
+			err := q.auth.Authorize(ctx, act, policy.ActionUpdate, rbacObj)
+			if err == nil {
```

Comment:
Add a comment that failures will always skip the cache and hit the database in case the cache is stale.